### ISSUE TITLE
Duplicate completion candidates

### ISSUE THREAD

**Issue by romkatv** [Aloxaf/fzf-tab#19]
To reproduce:

1. Install zsh 5.7.1 (I haven't tried any other version).
2. Run `git clone https://github.com/Aloxaf/fzf-tab.git ~/fzf-tab`.
3. Run `zsh -df`.
4. Run `autoload compinit && compinit`.
5. Run `cd ~/fzf-tab`.
6. Type `git checkout maste` and hit `TAB`.
7. Observe `maste` being completed to `master`.
8. Hit Ctrl-C and run `source ~/fzf-tab/fzf-tab.zsh`.
9. Type `git checkout maste` and hit `TAB`.
10. Observe the **unexpected** appearance of fzf:
```text
adam% git checkout maste
> master                                                                              
  2/2                                                                                 
> master                                                                              
  master  -- Minor cleanup (#17)                                                      
```

**Comment by romkatv** [comment]
It's fairly easy to fix this specific case. Basically, don't show fzf if all words in `compcap` are identical.

It's trickier to deal with duplicates in general. For example, there are 3 completion candidates for `git checkout m`:

- Without `fzf-tab` with `zstyle ':completion:*:descriptions' format ' -- %d --'`:
```text
adam% git checkout m
 -- recent branches --
master  -- Minor cleanup (#17)
 -- local head --
 -- head --
 -- remote branch name --
master        multi-select
```
- With `fzf-tab`:
```
adam% git checkout m
> m
  3/3
  master
  master  -- Minor cleanup (#17)
> multi-select
```

It sort of makes sense that `master` is listed twice in the first case: once in "recent branches" section and again in "remote branch name". It makes less sense in fzf because section names have disappeared.

I don't know what's the best solution here. Some ideas:

- Remove all duplicates. Which description should be kept then?
- If the same completion candidate is listed more than once, show section name in the description. Here's how it could look:
```
adam% git checkout m
> m
  3/3
  master        -- [remote branch name]
  master        -- [recent branches] Minor cleanup (#17)
> multi-select  -- [remote branch name]
```

**Comment by Aloxaf** [comment]
I prefer showing section name in the description. But it may cause the description to be too long......

**Comment by romkatv** [comment]
Yeah, formatting can be tricky.

What do you think about first fixing the simple case where all completion candidates are identical, and then think what can be done about duplicates in general? The simple case is the more disruptive and also easy to fix.

**Comment by Aloxaf** [comment]
I think that's a good idea. I'm considering if I can do this check when calculating the longest common prefix, or write an individual function.

**Comment by romkatv** [comment]
I sent you #20.

**Comment by Aloxaf** [comment]
Merged. Thank you very much.

**Comment by Aloxaf** [comment]
After reading the fzf's manual, I noticed that fzf has a `--header-lines` argument. It can let the first N lines of the input are treated as the sticky header.

Maybe I can use this to simplify the description, just like:
```
adam% git checkout m
> m
  3/3
  [a] remote branch name <- this two line are sticky headers
  [b] recent branches   <- this two line are sticky headers
> master        -- [a]
  master        -- [b] Minor cleanup (#17)
  multi-select  -- [a]
```

or more colorful with the help of `--ansi`
```
adam% git checkout m
> m
  3/3
  -- remote branch name -- <- this line has red foreground color
  -- recent branches -- <- this line has green foreground color
> master  <- red foreground color
  master        Minor cleanup (#17) <- green foreground color
  multi-select  <- red foreground color
```

This might be a better way to indicate the section.

**Comment by romkatv** [comment]
That sounds promising. It might be tricky to chose enough colors that are all readable and distinguishable. What if there are 20 sections?

Related: Use different colors for different kinds of files, following LS_COLORS or the equivalent style from zsh completion system. Easier with the file-specific fzf completer (atl-c, I believe, in the default fzf installation.)

**Comment by Aloxaf** [comment]
Hmm, I hope this won't happen. I think the only way to get enough distinguishable colors is let users choose them by themselves. Of course a good default value is also necessary.

Thanks for your tips.

**Comment by Aloxaf** [comment]
The basic work has been finished. ( #23 )
![图片](https://user-images.githubusercontent.com/17017672/72069197-9e3da980-3321-11ea-9efd-928ac743cf3d.png)

![图片](https://user-images.githubusercontent.com/17017672/72069239-b31a3d00-3321-11ea-9955-4ea76319547f.png)

**Comment by romkatv** [comment]
Super cool!

Those sections take quite a bit of space at the top. Is it possible to move them to the bottom? Maybeusing just one line for all sections?

```text
[recent-branches] [local head] ...
```

If this becomes longer than `$COLUMNS`, go to the next line and perhaps align them.

```text
[recent-branches] [local head        ]
[remote head    ] [remote branch name]
```

**Comment by Aloxaf** [comment]
It seems to impossible, fzf doesn't support put header at the bottom (after all it is header...). An approch is using `--preview` but it doesn't support flexable height ( must be a fixed value).

Put multiple headers in one line sounds great.

**Comment by Aloxaf** [comment]
`FZF_TAB_MERGE_HEADERS` is introduced in https://github.com/Aloxaf/fzf-tab/pull/23/commits/bbaf2be8f22f7531e1b006c0a44b1d64c9e4579a.

It uses tab for alignment, so it is quite ugly when tab width is small but the section name is long.
![图片](https://user-images.githubusercontent.com/17017672/72081356-4dd24600-3339-11ea-836c-ba4923009af4.png)

To solve this, users have to set tab width to a bigger value (eg. 20; default is 8):
![图片](https://user-images.githubusercontent.com/17017672/72081463-7ce8b780-3339-11ea-9f2f-b4bd2bfeeab1.png)

But this may waste some space when headers are all short.

Maybe I should allow the user to set the tab width separately for certain commands? 
eg. `zstyle ':completion:fzf_tab:git:' tab-width 20`

**Comment by romkatv** [comment]
That doesn't look very nice. What do you think about the following approach. First iterate over all elements in `_fzf_tab_groups` to compute the maximum length among all groups. This will give us column width for every group. The number of columns per line can be obtained by dividing `$COLUMNS` (this is terminal window width) by the number obtained above.

The output also looks rather crowded. Perhaps only those groups should be included for which there is more than one completion candidate with the same word?

**Comment by Aloxaf** [comment]
Seems great!
![image](https://user-images.githubusercontent.com/17017672/72130595-fbcf0600-33b4-11ea-8f5a-5e9b71420aa3.png)
![image](https://user-images.githubusercontent.com/17017672/72130603-038eaa80-33b5-11ea-9ff2-1ea0577d4400.png)


That sounds good. But some users (like me) would like to see full group list. I think it can be a optional feature.

**Comment by romkatv** [comment]
Awesome stuff!

I wonder if it'll look better without square brackets. Since all headers are aligned and printed with different colors, the structure should still be apparent even without brackets.

And +1 for some kind of option to control groups. If you decide to introduce it, it's probably a good idea to make its value open-ended rather than a boolean. Reasonable group handling strategies:

1. always show all groups
2. only show groups that are necessary to disambiguate between duplicate completion candidates
3. don't show groups but still show duplicate completion candidates
4. don't show groups and remove duplicate completion candidates (unclear which one to pick from the cluster of duplicates if they have different descriptions; the first, the last, random or the one with the longest description?)

There is also a room for another parameter that would control the format of completion option descriptions. It would allow users to achieve the following things:

1. Show "natural" descriptions.
2. Don't show option descriptions.
3. Prefix/suffix all option descriptions with their group names.
4. Prefix/suffix descriptions with their group names only for duplicate options.

If the description pattern is flexible enough, it could also allow the user to control colors or lack thereof.

These are just random ideas, please don't treat them as feature requests. You've been doing fantastic job making product decisions, so I trust you to make good choices here.

**Comment by Aloxaf** [comment]
The format of group name is controlled by `zstyle ':completion:*:descriptions' format` ( for me is  `zstyle ':completion:*:descriptions' format '[%d]'`, so there is square brackets). 
If not set no group name will be printed. ( This is consistent with the default behavior of zsh )

Thanks for you advice! They're helpful, I'll take them into consideration.

**Comment by Aloxaf** [comment]
The final version is basically complete.
![图片](https://user-images.githubusercontent.com/17017672/72215686-62892680-3551-11ea-9715-fc7c67f0b5dc.png)


For those who don't set `zstyle ':completion:*:descriptions' format`, they won't feel any change. There are no groups and colors.

For those who set it, there will be groups and colors and a configurable prefix character to indicate the color. 

I noticed that zsh's builtin selection menu will always show all groups. 
So I decided not to remove any duplicate candidates. But users can still choose to only show groups that have duplicate members. 
![图片](https://user-images.githubusercontent.com/17017672/72215556-c3632f80-354e-11ea-8f11-ea38545bed02.png)


Plase feel free to let me know if you have any suggesstions.
